File: code/SWE2-Project/Animal Tracker/src/test/java/edu/westga/cs3211/animaltracker/model/ProjectModel/TestAddUser.java
package edu.westga.cs3211.animaltracker.model.ProjectModel;

import static org.junit.jupiter.api.Assertions.*;

import org.junit.jupiter.api.Test;

import edu.westga.cs3211.animaltracker.model.Project;
import edu.westga.cs3211.animaltracker.model.User;
import edu.westga.cs3211.animaltracker.model.Role;

/**
 * Tests for the addUser method in Project class.
 * @author mrocker1
 */
class TestAddUser {

    @Test
    void testAddNullUser() {
        Project project = new Project("Test Project");

        assertThrows(IllegalArgumentException.class, () -> project.addUser(null));
    }

    @Test
    void testAddValidUser() {
        Project project = new Project("Test Project");
        User user = new User("testuser", "password123", Role.SCIENTIST);

        project.addUser(user);

        assertEquals(1, project.getUsers().size(), "checking users list size");
        assertTrue(project.getUsers().contains(user), "checking user is in list");
    }

    @Test
    void testAddMultipleUsers() {
        Project project = new Project("Test Project");
        User user1 = new User("user1", "password1", Role.SCIENTIST);
        User user2 = new User("user2", "password2", Role.CONTRIBUTOR);
        User user3 = new User("user3", "password3", Role.GUEST);

        project.addUser(user1);
        project.addUser(user2);
        project.addUser(user3);

        assertEquals(3, project.getUsers().size(), "checking users list size");
        assertTrue(project.getUsers().contains(user1), "checking user1 is in list");
        assertTrue(project.getUsers().contains(user2), "checking user2 is in list");
        assertTrue(project.getUsers().contains(user3), "checking user3 is in list");
    }

    @Test
    void testAddDuplicateUser() {
        Project project = new Project("Test Project");
        User user = new User("duplicate", "password", Role.SCIENTIST);

        project.addUser(user);

        assertThrows(IllegalArgumentException.class, () -> project.addUser(user));
    }

    @Test
    void testAddUserToProjectWithExistingUsers() {
        Project project = new Project("Test Project");
        User user1 = new User("existing", "pass1", Role.SCIENTIST);
        User user2 = new User("newuser", "pass2", Role.ADMIN);

        project.addUser(user1);
        int initialSize = project.getUsers().size();

        project.addUser(user2);

        assertEquals(initialSize + 1, project.getUsers().size(), "checking size increased by 1");
        assertTrue(project.containsUser(user2), "checking new user exists");
    }
}

File: code/SWE2-Project/Animal Tracker Server/src/test/model_test/project_test/TestAddUser.py
import unittest

from model.Project import Project
from model.Role import Role
from model.User import User


class TestAddUser(unittest.TestCase):
    def test_add_one_user(self):
        admin = User("Tom", "1234", Role.ADMIN)
        project = Project("1234", [], [], 1)
        project.add_user(admin)
        returned_users = project.get_users()
        self.assertEqual(len(returned_users), 1)
        self.assertEqual(admin, returned_users[0])

    def test_add_multiple_user(self):
        admin = User("Tom", "1234", Role.ADMIN)
        scientist = User("Bob", "1234", Role.SCIENTIST)
        project = Project("test", [], [], 1)
        project.add_user(admin)
        project.add_user(scientist)
        returned_users = project.get_users()
        self.assertEqual(len(returned_users), 2)
        self.assertTrue(scientist in returned_users)
        self.assertTrue(scientist in returned_users)

    def test_add_none_user(self):
        project = Project("test", [], [], 1)
        with self.assertRaises(Exception):
            project.add_user(None)

    def test_add_duplicate_user(self):
        project = Project("test", [], [], 1)
        admin = User("Tom", "1234", Role.ADMIN)
        project.add_user(admin)
        with self.assertRaises(Exception):
            project.add_user(admin)

